File: summary/python/14_greedy.py
def greedy_algorithm(values, weights, capacity):
    items = sorted(
        [(i, v, w, v/w) for i, (v, w) in enumerate(zip(values, weights), 1)],
        key=lambda x: x[3],
        reverse=True
    )

    total_value = 0
    total_weight = 0
    selected_items = []

    for index, value, weight, _ in items:
        if total_weight + weight <= capacity:
            selected_items.append(index)
            total_value += value
            total_weight += weight

    return total_value, selected_items

# Example usage:
values = [60, 100, 120]
weights = [10, 20, 30]
capacity = 50
value, items = greedy_algorithm(values, weights, capacity)
print(f"Total value: {value}")
print(f"Items selected: {', '.join(map(str, items))}")

File: summary/elixir/14_greedy.exs
defmodule GreedyAlgorithm do
  def knapsack(values, weights, capacity) do
    items = Enum.zip(values, weights)
    |> Enum.with_index(1)
    |> Enum.map(fn {{value, weight}, index} ->
      %{index: index, value: value, weight: weight, ratio: value / weight}
    end)
    |> Enum.sort_by(&(&1.ratio), :desc)

    {total_value, total_weight, selected_items} =
      Enum.reduce_while(items, {0, 0, []}, fn item, {value, weight, selected} ->
        if weight + item.weight <= capacity do
          {:cont, {value + item.value, weight + item.weight, [item.index | selected]}}
        else
          {:halt, {value, weight, selected}}
        end
      end)

    {total_value, Enum.reverse(selected_items)}
  end
end

# 使用例
values = [60, 100, 120]
weights = [10, 20, 30]
capacity = 50

{value, items} = GreedyAlgorithm.knapsack(values, weights, capacity)
IO.puts "Total value: #{value}"
IO.puts "Items selected: #{Enum.join(items, ", ")}"
